### DISCUSSION TITLE
Script to Create Notes from Annotations

### DISCUSSION THREAD

**Discussion by Bulgy404** [windingwind/zotero-actions-tags#275] [Q&A]
Hey guys, 

glad I stumbled upon this git :) I have a Zotero Library with about 2500 Papers which have all been marked in the corresponding pdf. Now I connected Zotero to my Notion and have a nice overview of all the papers. For Notion to read the markings in them I need to "Add note from annotation" for every paper. Doing this manually is kind of a hassle tbh. 

Has anybody an idea how I could automate this? that would be amazing 👍 

Best
![Screenshot_Zotero_2024-03-01@2x](https://github.com/windingwind/zotero-actions-tags/assets/115490923/3933419c-bddc-4f36-9811-9d12160d70c5)

**Comment by ulahcherubim** [comment]
Well, I have managed to write a script that works when triggered by a shortcut or close tab. But it doesn't overwrite previously created notes from annotations, which is a problem because everytime you open and close a PDF you will get another note under your item. I don't want to write anything that deletes anything  because it might harm in the long run. So I will provide you with a script that creates a note from annotations. 

Event: Close Tab(if you do that it creates note from annotations every time you close a pdf) or None 
Action: Script
Data:
```
/**
 * Creates Note From Annotations  .
 * @author ulahcherubim
 * @usage 
 * @link 
 * @see https://github.com/windingwind/zotero-actions-tags/discussions/275
 */
const ZoteroPane = require("ZoteroPane");
ZoteroPane.addNoteFromAnnotationsFromSelected();

```
If anyone can add how to decide an item has already a child note or not and whether you added an annotation or not (which are queries for api i guess) , this would work well. Otherwise, you will have so many annotation notes.

**Reply by Bulgy404** [reply]
Hey, thanks for your contribution :) I found myself another way of managing the papers which might be interesting for others as well.

Becaue Notero has some limitations to read those annotations as described above I use Obsidian now to sort the annotations and manage them and use notion to tag my papers and move them into different groups. This works really well as you can embedd an obsidian link in your notion database to the corresponding papers. If anybody wants more information on that feel free to reach out.

Cheers

**Reply by oliverkuehle** [reply]
Hey, does that mean you retrieve the annotations in Obsidian directly from the pdf? So without the need of "Add note from Annotations"? Could you kindly share the plugin that you chose. Thanks

**Reply by Bulgy404** [reply]
Hey @oliverkuehle 

the add-on for Obsidian is called "Zotero Integration" you can find it in the Obsidian Store (https://github.com/mgmeyers/obsidian-zotero-integration)

I am also using this template for the annotation
`---

tags: [{{allTags}}]
title: "{{title}}"
authors: "{{authors}}"
year: {{date | format("YYYY")}}
publisher: "{{publicationTitle}}"
citekey: {{citekey}}
collection: "[[{{collections[0].fullPath}}]]"

---

> [!meta]- Metadata
> abstract:: {{abstractNote}}
> zotero_link:: {{pdfZoteroLink}}
> Related:: {% for relation in relations -%} {%- if relation.citekey -%} [[{{relation.citekey}}]], {% endif -%} {%- endfor%}
> url:: {{url}}
> doi:: {{doi}}
> bibliography:: {{bibliography}}

---

  

### Webpage

  

<iframe src="{{url}}" allow="fullscreen" allowfullscreen="" style="height:100%;width:100%; aspect-ratio: 16 / 10; "></iframe>

  

---

  

## Self Notes 

{% persist "notes" %}

  
  
  

{% endpersist %}

---

## Reading notes

  

{%-

set zoteroColors = {

"#2ea8e5": "blue",

"#5fb236": "green",

"#a28ae5": "purple",

"#ff6666": "red",

"#ffd400": "yellow",

"#f19837": "orange",

"#aaaaaa": "grey",

"#e56eee": "magenta"

}

-%}

  

{%-

set colorHeading = {

"blue": "ℹ Background information, Prerequisites",

"green": "❓ Assumptions, Questions, Goals, Problems",

"purple": "📊 Main findings, Results, Conclusions",

"red": "🧪Experimental details or Methods",

"yellow": "⭐ Interesting point, Facts, Examples",

"orange": "⚠️ Discussion, Disagree with author",

"grey": "📅 Vocabulary, Names, Dates, Definitions",

"magenta": "📄 Important references"

}

-%}

  

{%- macro calloutHeader(type) -%}

{%- switch type -%}

{%- case "highlight" -%}

Highlight

{%- case "image" -%}

Image

{%- default -%}

Note

{%- endswitch -%}

{%- endmacro %}

  

{%- set newAnnotations = [] -%}

  

{% if annotations.length > 0 %}

  

{%- for annot in annotations -%}

  

{%- if annot.color in zoteroColors -%}

{%- set customColor = zoteroColors[annot.color] -%}

{%- elif annot.colorCategory|lower in colorHeading -%}

{%- set customColor = annot.colorCategory|lower -%}

{%- else -%}

{%- set customColor = "other" -%}

{%- endif -%}

  

{%- set newAnnotations = (newAnnotations.push({"annotation": annot, "customColor": customColor}), newAnnotations) -%}

  

{%- endfor -%}

  

{#- INSERT ANNOTATIONS -#}

{#- Loops through each of the available colors and only inserts matching annotations -#}

{#- This is a workaround for inserting categories in a predefined order (instead of using groupby & the order in which they appear in the PDF) -#}

  

{%- for color, heading in colorHeading -%}

{%- for entry in newAnnotations | filterby ("customColor", "startswith", color) -%}

{%- set annot = entry.annotation -%}

  

{%- if entry and loop.first %}

  

### [[{{colorHeading[color]}}]]

{%- endif %}

  

> [!quote{{"|" + color if color != "other"}}]+ {{calloutHeader(annot.type)}} ([page. {{annot.pageLabel}}](zotero://open-pdf/library/items/{{annot.attachment.itemKey}}?page={{annot.pageLabel}}&annotation={{annot.id}}))

  

{%- if annot.annotatedText %}

> {{annot.annotatedText|nl2br}} {% if annot.hashTags %}{{annot.hashTags}}{% endif -%}

{%- endif %}

  

{%- if annot.imageRelativePath %}

> ![[{{annot.imageRelativePath}}]]

{%- endif %}

  

{%- if annot.ocrText %}

> {{annot.ocrText}}

{%- endif %}

  

{%- if annot.comment %}

> - **{{annot.comment|nl2br}}**

{%- endif -%}

{%- endfor -%}
{%- endfor -%}
{% endif %}`


I hope this help, Best

**Comment by oliverkuehle** [comment]
Building on ulahcherubim's comment I added some code to the script. 
It first deletes all existing _Extracted Annotation_ notes and then creates the new one.

- Like @ulahcherubim mentioned, it works with the action event set to 'Close Tab'.
- However if you add and run the action using menus, it will the script will run for every selected zotero entry. For example, you can highlight your whole library and then click the action, which will update "Extracted Annotation" notes for the whole library. This way @Bulgy404 can update all 2500 at once :) 

```const ZoteroPane = require("ZoteroPane");
const items = await ZoteroPane.getSelectedItems()
for (item of items) {
    const noteIDs = item.getNotes();
    for (let id of noteIDs) {
        let note = Zotero.Items.get(id);
        let noteHTML = note.getNote();
        if (noteHTML.startsWith('<div data-citation-items') && noteHTML.includes('Annotations')) {
            Zotero.Items.trashTx(id);   
        }
    }       
}
ZoteroPane.addNoteFromAnnotationsFromSelected();
```

_Note: If you're using Zotero in a different language than english, you will have to replace the 'Annotations' in line 8 with whatever title stand at the start of your extracted annotation notes._

**Be aware that this works for me, but might not work for you or might have become deprecated by the time you read this. Use at your own risk, as there is no guarantee that it will work correctly.**

**Comment by EnriqueSolarte** [comment]
I managed to create a script that createsa  note from an annotation and deletesthe  previous note
```
if (items?.length > 0) {
    for (let item of items) {
        // Delete existing annotation notes
        const noteIDs = item.getNotes();
        for (let id of noteIDs) {
            let note = Zotero.Items.get(id);
            let noteHTML = note.getNote();
            if (noteHTML.startsWith('<div data-citation-items') && noteHTML.includes('Annotations')) {
                Zotero.Items.trashTx(id);   
            }
        }
    }
}
const ZoteroPane = require("ZoteroPane");
await ZoteroPane.addNoteFromAnnotationsFromSelected();

```

**Reply by EnriqueSolarte** [reply]
This is a better script. It first verifies that the parent Item is selected (sometimes, I executed this script from the Panel reader). Then, it deletes all note that has the word "Annotation" in the title. Finally, create a new note from Annotations. and add a tag 'ready' to all items.
```
const ZoteroPane = require("ZoteroPane");

// Attempt to retrieve selected items
items = ZoteroPane.getSelectedItems();

// If no items are selected, check if we're in the PDF reader tab
if (!items || items.length === 0) {
    const zp = Zotero.getActiveZoteroPane();
    const currentTab = zp.getCurrentTab();

    if (currentTab && currentTab.type === 'reader') {
        const pdfItem = currentTab.item;
        if (pdfItem && pdfItem.isAttachment()) {
            const parentItem = pdfItem.getSource();
            if (parentItem) {
                items = [parentItem];
            }
        }
    }
}

// Proceed if we have items to process
if (items && items.length > 0) {
    for (let item of items) {
		item.addTag('ready');
        // Delete existing annotation notes
		if (item.isNote()) {
            continue;
        }
        const noteIDs = item.getNotes();
        for (let id of noteIDs) {
            const note = Zotero.Items.get(id);
            if (note.isNote()) {
                const noteHTML = note.getNote();
                if (noteHTML.startsWith('<div data-citation-items') && noteHTML.includes('Annotations')) {
                    Zotero.Items.trashTx(id);
                }
            }
        }
    }

    // Create a new note from annotations
    await ZoteroPane.addNoteFromAnnotationsFromSelected();
}

```
